lib/cache/cache.go

```diff
@@ -388,18 +533,43 @@ func New(config Config) (*Cache, error) {
 // to handle subscribers connected to the in-memory caches
 // instead of reading from the backend.
 func (c *Cache) NewWatcher(ctx context.Context, watch services.Watch) (services.Watcher, error) {
+Outer:
+	for _, requested := range watch.Kinds {
+		for _, configured := range c.Config.Watches {
+			if requested.Kind == configured.Kind {
+				continue Outer
+			}
+		}
+		return nil, trace.BadParameter("cache %q does not support watching resource %q", c.Config.target, requested.Kind)
+	}
 	return c.eventsFanout.NewWatcher(ctx, watch)
 }
 
 func (c *Cache) isClosed() bool {
 	return atomic.LoadInt32(&c.closedFlag) == 1
 }
 
+func (c *Cache) isClosing() bool {
+	if c.isClosed() {
+		return true
+	}
+	select {
+	case <-c.ctx.Done():
+		return true
+	default:
+		return false
+	}
+}
+
 func (c *Cache) setClosed() {
 	atomic.StoreInt32(&c.closedFlag, 1)
 }
 
 func (c *Cache) update(ctx context.Context) {
+	defer func() {
+		c.Debugf("Cache is closing, returning from update loop.")
+		c.Close()
+	}()
```

Comment:
not sure how the cache will know if the retry impl below fails to construct - why is `retry` not an argument for `update`?